### DISCUSSION TITLE
2 patterns, 1 of those 2 can be at different places, both with 2 different colors, and both can fade

### DISCUSSION THREAD

**Discussion by DynV** [ImageMagick/ImageMagick#7464] [Help]
### ImageMagick version

7.1.1-34 Q16-HDRI x64 b0b7b17:20240623

### Operating system, version and so on

Windows 10 22H2

### Description

A non-negligible portion of screenshots are of my leveling up. There are 2 indicators of these at different places, the 1st is always at the same place starting with "LEV", the 2nd is always at the same horizontal position, but the vertical one can change, starting with "You've reached level". Both indicators fade at different rates and have the same 2 different possible base colors, green and beige. As you can see in the following the 1st pattern is ~1/6 horizontally and ~1/2 vertically, the 2nd one is ~1/20 horizontally and vertically usually at the top but could be as low as right above the 1st indicator. Notice the 2nd indicator is always accompanied with a non-faded color line to the left.
![20240701204108_1](https://github.com/ImageMagick/ImageMagick/assets/2823807/b093f640-96b1-4250-83c9-6e0f148a6a99)
![20240707040157_1](https://github.com/ImageMagick/ImageMagick/assets/2823807/e0d384d7-ac33-48c9-9082-4ccdd54909a2)

Is there a way to list all, or most, leveling up screenshots?

Thank you kindly

**Comment by snibgo** [comment]
If I understand you correctly, then a crop at 144x107+561+923 should be readable as the characters "LEV". Windows BAT script:

```
set INFILE=%1
if "%INFILE%"=="" set INFILE=dynv1.jpg

%IMG7%magick %INFILE% ^
  -crop 144x107+561+923 +repage -resize 20x20 ^
  -colorspace Gray -negate ^
  x.png

%TESS% x.png dynv

type dynv.txt
```

%TESS% runs Tesseract.

For both your inputs, the output text file contains "LEV".

**Reply by DynV** [reply]
They fade at different rate. I try to take the screenshot as soon as possible but sometimes 1 has completely disappeared and the remaining is fading out. Thus I'd like both ways to be checked.

**Comment by fmw42** [comment]
For the second region, if the font size is always the same, then you could crop out "You've reached level" and do template matching using magick compare to find that text in your images.

**Reply by DynV** [reply]
`the font size is always the same`
Indeed. How would I do that and is there a way for both methods to be checked at the same time? That way the files don't need to be read 2 times.

**Comment by fmw42** [comment]
No way that I know to do it all in one command.  You know the first crop location from snibgo's comment above. The second one you can find by making a grayscale template by cropping image1 around the text you need to find.  I have done that. That gives you the size of the second crop you want. Then you need to use magick compare to find the location to crop. Once you have that, you can do both crops at the same time.

`crop the first region at  at 144x107+561+923
`

Template:

![template](https://github.com/ImageMagick/ImageMagick/assets/13458218/69fb315b-2fbc-47a7-960e-6c9497cd4ff4)


```
Get template size (one time only):

magick template.png -format "%wx%h" info:
443x50
```

```
Do the compare to find the match score and location of the template within each image:

magick compare image1.jpg template.png -colorspace gray -metric NCC -subimage-search null:
0.999944 @ 221,256

magick compare image2.jpg template.png -colorspace gray -metric NCC -subimage-search null:
0.960823 @ 221,256
```

```
Now crop both sections from each image:

magick image1.jpg -write mpr:img +delete ^
( mpr:img -crop 144x107+561+923 +repage +write image1_sub1.jpg +delete ) ^
( mpr:img -crop 443x50+221+256 +repage +write image1_sub2.jpg +delete ) ^
null:

magick image2.jpg -write mpr:img +delete ^
( mpr:img -crop 144x107+561+923 +repage +write image2_sub1.jpg +delete ) ^
( mpr:img -crop 443x50+221+256 +repage +write image2_sub2.jpg +delete ) ^
null:
```

One of the Windows users will have to show you how to get the match location into variables to use with the crop commands. I am not a Windows user, sorry.

**Reply by DynV** [reply]
The indicator starting with "You've reached level" is at different vertical positions. Sorry if the OP wasn't clear enough about that.

**Reply by fmw42** [reply]
In the two images you posted, it happens to be at the same vertical location.  The code I presented with show different vertical positions if that is the case.  So you would use that value when cropping.  Some Windows user will need to show you how to get those values as variables from the compare and then use them in the crop.

**Reply by fmw42** [reply]
Here are the two crop images for each image from my code. They look correct to me.

![image1_sub1](https://github.com/ImageMagick/ImageMagick/assets/13458218/f6b1d2ee-506a-4f2a-9158-40cc032d3d00)

![image1_sub2](https://github.com/ImageMagick/ImageMagick/assets/13458218/b81bba3e-2f2f-423b-ad88-0e7b3ab6c2d4)

![image2_sub1](https://github.com/ImageMagick/ImageMagick/assets/13458218/cbb9edb4-efaf-459a-a354-1f93eed7f66f)

![image2_sub2](https://github.com/ImageMagick/ImageMagick/assets/13458218/9020264b-2b14-4e9e-85a3-c86888669b6d)

**Reply by DynV** [reply]
For these 2 cases they're the same but I--assure--you they can vary.

**Comment by fmw42** [comment]
And my code will adjust to find them.  That is what magick compare does.  It searches the large image to find the best match for the small image.

**Comment by snibgo** [comment]
For performance of the search for "You've reached level", I suggest cropping to the area that can contain that text, and searching just that area.

> ... and is there a way for both methods to be checked at the same time? That way the files don't need to be read 2 times.

"You've reached level" can occur at a variable position, so we need to use `-subimage-search` to find it. That option is available only in `magick compare`. So we need at least one `magick compare` command. We could do multiple crops within that command.

But each file doesn't need to be read 2 times. The first `magick` command can crop out the only position that "LEV" can occur, and also crop out the larger area that can contain "You've reached level". Then the `magick` compare need read only the area that can contain "You've reached level".

**Reply by fmw42** [reply]
But then you need another magick command to crop the second image found by magick compare, if I understood the OP correctly.

**Reply by DynV** [reply]
My issue is to having to read the image twice, not having to run a command twice (or 2 one time).

**Reply by fmw42** [reply]
No way that I know to do that.  Magick and Magick compare are two different tools and cannot share an image.

**Reply by DynV** [reply]
It was a preference, not a requirement, I'll do it that way.

**Reply by fmw42** [reply]
Python/OpenCV/Numpy can do what you want. But you will need to learn those tools.  It has a very fast template matching method and Numpy can do the cropping very quickly also. Once an image is loaded, it is in memory until the script ends.
